Fixing uncaught exception problem on shutdown

do3cc opened the pull request:
Nested connections try to clean up themselves multiple
times. This does not work.

I'd like to have a review from somebody, I don't feel 1000 % confident about the change.

do3cc commented:
#45

tseaver commented on src/CHANGES.txt:
```diff
@@ -18,6 +18,8 @@ Bugs Fixed
 - Pinned the ``transaction`` and ``manuel`` dependencies to Python 2.5-
   compatible versions when installing under Python 2.5.
 
+- Cleanup of nested mvcc connections could have failed.
```
The convention is to have the changelog entries in imperative mode.  Something like:

```
- Avoid failure due to repeated cleanup of a connection's storage (typically during shutdown).
```

tseaver commented:
Aside from the changelog issue, LGTM.

mgedmin commented:
How hard would it be to write a unit test for this?

do3cc commented:
@mgedmin very hard. There comes the I am not 1000% sure what I am doing into play. But I am trying.

do3cc commented on src/ZODB/tests/testMVCCMappingStorage.py:
```diff
@@ -33,6 +33,15 @@
+        db2 = DB(None, databases=db1.databases, database_name='2')
+        db1.open().get_connection('2')
+        db1.close()
+        db2.close()
```
Zope stores all dbs that have been created. On shutdown, it closes each of them, I am mirroring this here.

do3cc commented on src/ZODB/tests/testMVCCMappingStorage.py:
```diff
@@ -33,6 +33,15 @@
+        # https://github.com/zopefoundation/ZODB/issues/45
+        db1 = DB(self._storage)
+        db2 = DB(None, databases=db1.databases, database_name='2')
+        db1.open().get_connection('2')
```
I must admin, I am only triggering observed behavior. From my understanding, the information that db1 has somehow a connection to db2 is lazy added when I try to get the second connection from it.

do3cc commented on src/ZODB/Connection.py:
```diff
@@ -1073,7 +1073,8 @@ def _resetCache(self):
             if c._mvcc_storage:
-                c._storage.release()
+                if c._storage is not None:
+                    c._storage.release()
```
When I wrote the fix, I compared a stock Plone page with our page with Relstorage.
The reason why this causes no exception in a stock Plone/Zope page is that the storage used does not provide mvcc_capability, this is why c._mvcc_storage is False. A stock FileStorage depends on the connection to provide MVCC (From: https://github.com/zopefoundation/ZODB/blob/master/src/ZODB/interfaces.py#L1022)

Since Relstorage supports MVCC at storage level, this code finally get executed and fails with multiple, nested databases. Which you already have with a stock Zope Site, because of the tempstorage.

do3cc commented:
@mgedmin I got a working test that triggers an error without my fix
@tseaver I've updated the changelog.

I've added comments to my test and my change in github about my understanding of what I am doing. I hope I got it all right.

tseaver commented:
@do3cc Thanks for your efforts!
